File: spatial-index/src/main/java/lva/spatialindex/index/RStarTree.java
package lva.spatialindex.index;

import lva.spatialindex.index.Distributions.GroupPair;
import lva.spatialindex.storage.Storage;

import java.awt.*;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.function.Function;
import java.util.function.Predicate;

import static lva.spatialindex.index.Distributions.getDistributions;
import static lva.spatialindex.index.Distributions.marginGroups;
import static lva.spatialindex.index.Entry.union;
import static lva.spatialindex.index.RStarTree.Utils.minList;
import static lva.spatialindex.index.Rectangles.area;


/**
 * @author vlitvinenko
 */
public class RStarTree implements Index {
    private final Storage<Node> storage;
    private Node root;

    public RStarTree(int maxNumberOfElements, String storageFileName) {
        long size = 64 * 1024L * 1024L; // TODO: make it as func (maxNumberOfElements)
        this.storage = new NodeStorage(storageFileName, size);
        this.root = Node.newNode(this.storage);
    }

    @Override
    public Collection<Long> search(Rectangle area) {
        return search(root, area);
    }

    private Collection<Long> search(Node node, Rectangle area) {
        Collection<Long> res = new HashSet<>();
        for (Entry e: node.getEntries()) {
            if (area.intersects(e.getMbr())) {
                if (node.isLeaf()) {
                    res.add(-(e.getChildOffset() + 1));
                } else {
                    Collection<Long> subRes = e.getChildNode()
                            .map(childNode -> search(childNode, area))
                            .orElse(Collections.emptyList());
                    res.addAll(subRes);
                }
            }
        }
        return res;
    }

    public void insert(long offset, Rectangle newMbr) {
        if (!newMbr.isEmpty()) {
            insert(root, offset, newMbr);
        }
    }

    private void insert(Node node, long offset, Rectangle newMbr) {
        Node leafNode = chooseSubtree(node, newMbr);
        Node newNode = null;

        Entry entry = new Entry(storage, newMbr, -(offset + 1));
        if (!leafNode.isFull()) {
            leafNode.addEntry(entry);
        } else {
            newNode = splitNode(leafNode, Node.newNode(storage).addEntry(entry));
        }

        adjust(leafNode, newNode);
    }

    private static Node chooseSubtree(Node node, Rectangle newMbr) {
        if (node.isLeaf()) {
            return node;
        }

        Function<Entry, Long> bySize = e -> area(e.getMbr());
        Function<Entry, Long> byEnlarge = e -> area(e.getMbr().union(newMbr));
        Function<Entry, Long> byCost = e -> area(e.getMbr().intersection(newMbr));

        List<Entry> candidates = node.getEntries();

        // check that points to leaf node
        boolean isContainsLeaf = node.getEntries().stream().findAny()
                .flatMap(Entry::getChildNode)
                .map(Node::isLeaf).orElse(false);

        if (isContainsLeaf) {
            candidates = minList(candidates, byCost);
        }

        candidates = minList(candidates, byEnlarge);
        candidates = minList(candidates, bySize);

        return candidates.stream().findAny().flatMap(Entry::getChildNode)
                .map(childNode -> chooseSubtree(childNode, newMbr))
                .orElse(node);
    }

    private static Node splitNode(Node node, Node newNode) {
        if (node.isLeaf() != newNode.isLeaf()) {
            return newNode;
        }

        // distribute entries between nodes

        List<Entry> allEntries = new ArrayList<>(node.getEntries().size() + newNode.getEntries().size());
        allEntries.addAll(node.getEntries());
        allEntries.addAll(newNode.getEntries());

        // sort by x axis
        List<Entry> sort1ByX = new ArrayList<>(allEntries);
        sort1ByX.sort(Entry.LEFT_TO_RIGHT_BY_LEFT_COMPARATOR);
        List<Entry> sort2ByX = new ArrayList<>(allEntries);
        sort2ByX.sort(Entry.LEFT_TO_RIGHT_BY_RIGHT_COMPARATOR);

        // sort by y axis
        List<Entry> sort1ByY = new ArrayList<>(allEntries);
        sort1ByY.sort(Entry.TOP_TO_BOTTOM_BY_TOP_COMPARATOR);
        List<Entry> sort2ByY = new ArrayList<>(allEntries);
        sort2ByY.sort(Entry.TOP_TO_BOTTOM_BY_BOTTOM_COMPARATOR);

        List<GroupPair> groups1X = getDistributions(sort1ByX);
        List<GroupPair> groups2X = getDistributions(sort2ByX);
        List<GroupPair> groups1Y = getDistributions(sort1ByY);
        List<GroupPair> groups2Y = getDistributions(sort2ByY);

        int marginX = marginGroups(groups1X) + marginGroups(groups2X);
        int marginY = marginGroups(groups1Y) + marginGroups(groups2Y);

        List<GroupPair> axisGroups = new ArrayList<>();
        axisGroups.addAll(marginX < marginY ? groups1X : groups1Y);
        axisGroups.addAll(marginX < marginY ? groups2X : groups2Y);


        // find min overlapped values distribution
        List<GroupPair> candidatesOverlapped = minList(axisGroups, g -> {
            Rectangle gr1 = union(g.group1);
            Rectangle gr2 = union(g.group2);
            return area(gr1.intersection(gr2));
        });

        candidatesOverlapped = minList(candidatesOverlapped,
            g -> area(union(g.group1)) + area(union(g.group2)));

        GroupPair pair = candidatesOverlapped.isEmpty() ? new GroupPair()
            : candidatesOverlapped.get(0);

        node.setEntries(pair.group1);
        newNode.setEntries(pair.group2);

        return newNode;
    }

    private void adjust(Node node1, Node node2) {
        if (node1.getOffset() == root.getOffset()) {
            // node1 is root node
            if (node2 != null) {
                root = Node.newNode(storage)
                    .addNode(node1)
                    .addNode(node2);
            }
            return;
        }

        Node parent = storage.read(node1.getParentOffset());
        syncMbr(parent, node1);

        Node newNode = null;
        if (node2 != null) {
            if (!parent.isFull()) {
                parent.addNode(node2);
            } else {
                newNode = splitNode(parent, Node.newNode(storage).addNode(node2));
            }
        }

        node1 = parent;
        node2 = newNode;

        adjust(node1, node2);
    }

    private static void syncMbr(Node targetNode, Node sourceNode) {
        Predicate<Entry> isTargetEntry = entry -> entry.getChildNode()
                .map(Node::getOffset)
                .map(offset -> offset == sourceNode.getOffset()).orElse(false);

        Optional<Entry> targetEntry = targetNode.getEntries().stream()
                .filter(isTargetEntry).findAny();

        targetEntry.ifPresent(entry -> {
            entry.setMbr(sourceNode.getMbr());
            targetNode.resetMbr()
                    .save();
        });
    }

    @Override
    public void close() {
        storage.close();
    }

    static class Utils {
        private Utils() {}

        static <T> List<T> minList(List<? extends T> list, Function<? super T, Long> criteria) {
            List<T> candidates = new ArrayList<>();
            long minValue = Integer.MAX_VALUE;

            for (T e : list) {
                long ovrArea = criteria.apply(e);
                if (ovrArea <= minValue) {
                    if (ovrArea < minValue)  {
                        candidates.clear();
                        minValue = ovrArea;
                    }
                    candidates.add(e);
                }
            }

            return candidates;
        }
    }
}

File: spatial-index/src/main/kotlin/lva/spatialindex/index/RStarTree.kt
package lva.spatialindex.index

import arrow.core.getOrHandle
import lva.spatialindex.index.Node.Companion.newNode
import java.awt.Rectangle
import java.util.concurrent.locks.ReentrantReadWriteLock
import kotlin.concurrent.read
import kotlin.concurrent.write

/**
 * @author vlitvinenko
 */
class RStarTree(maxNumberOfElements: Int, storageFileName: String) : Index {
    private val storage = NodeStorage(storageFileName, SIZE_DEFAULT) // TODO: move out creation
    private val lock = ReentrantReadWriteLock()
    private var root: Node = newNode(storage)

    override fun search(area: Rectangle): Collection<Long> {
        lock.read {
            return search(root, area).toList()
        }
    }

    private fun search(node: Node, area: Rectangle): Sequence<Long> =
        node.entries.asSequence()
            .filter { entry ->
                area.intersects(entry.mbr)
            }
            .flatMap { entry ->
                entry.data()
                    .map { value -> sequenceOf(-(value + 1)) }
                    .getOrHandle { childNode -> search(childNode, area) }
            }

    fun insert(offset: Long, newMbr: Rectangle) {
        check(!newMbr.isEmpty) { "Invalid region" }
        lock.write {
            insert(root, offset, newMbr)
        }
    }

    private fun insert(node: Node, offset: Long, newMbr: Rectangle) {
        val leafNode = chooseSubtree(node, newMbr)
        val entry = Entry(storage, newMbr, -(offset + 1))
        val newNode = when {
            leafNode.isFull -> splitNode(leafNode, newNode(storage).addEntry(entry))
            else -> null.also { leafNode.addEntry(entry) }
        }

        adjust(leafNode, newNode)
    }

    private fun adjust(node1: Node, node2: Node?) {
        if (node1.offset == root.offset) {
            // node1 is root node
            if (node2 != null) {
                root = newNode(storage)
                    .addNode(node1).addNode(node2)
            }

        } else {

            val parent = storage.read(node1.parentOffset)
            syncMbr(parent, node1)

            var newNode: Node? = null
            if (node2 != null) {
                if (parent.isFull) {
                    newNode = splitNode(parent, newNode(storage).addNode(node2))
                } else {
                    parent.addNode(node2)
                }
            }

            adjust(parent, newNode)
        }
    }

    private fun chooseSubtree(node: Node, newMbr: Rectangle): Node {
        if (node.isLeaf) {
            return node
        }

        var candidates: List<Entry> = node.entries
        val isContainsLeaf = node.entries.firstOrNull()?.childNode?.isLeaf ?: false

        if (isContainsLeaf) {
            candidates = candidates.minList { it.mbr.intersection(newMbr).area } // by cost
        }

        candidates = candidates.minList { it.mbr.union(newMbr).area } // by enlarge
        candidates = candidates.minList { it.mbr.area } // by size

        return candidates.firstOrNull()?.childNode?.let { chooseSubtree(it, newMbr) } ?: node
    }

    private fun splitNode(node: Node, newNode: Node): Node {
        if (node.isLeaf != newNode.isLeaf) {
            return newNode
        }

        // distribute entries between nodes
        val allEntries = node.entries + newNode.entries

        val groupsX = getDistributionGroups(allEntries, listOf(Entry::left, Entry::right))
        val groupsY = getDistributionGroups(allEntries, listOf(Entry::bottom, Entry::top))
        val groups = if (getGroupMargins(groupsX) < getGroupMargins(groupsY))
            groupsX else groupsY

        // find min overlapped values distribution
        var overlapped = groups.flatten()
        overlapped = overlapped.minList { it.first.union().intersection(it.second.union()).area }
        overlapped = overlapped.minList { it.first.union().area + it.second.union().area }

        val pair = overlapped.firstOrNull() ?: GroupPair(arrayListOf(), arrayListOf())
        node.setEntries(pair.first)
        newNode.setEntries(pair.second)

        return newNode
    }

    private fun syncMbr(targetNode: Node, sourceNode: Node) {
        val targetEntry = targetNode.entries.firstOrNull { entry ->
            entry.childNode?.offset == sourceNode.offset
        }

        targetEntry?.let { entry ->
            entry.mbr = sourceNode.mbr
            targetNode.resetMbr().save()
        }
    }

    fun clear() = lock.write {
        root.reset()
        storage.clear()
    }

    companion object {
        private const val SIZE_DEFAULT = 64 * 1024 * 1024
    }
}

val Rectangle.area: Long get() = (width * height).toLong()
val Rectangle.margin: Int get() = 2 * (height + width)


private inline fun <T> List<T>.minList(criteria: (T) -> Long): List<T> {
    val min = minOfOrNull(criteria)
    return min?.let { filter { criteria(it) == min } } ?: arrayListOf()
}

